File: android-app/app/src/main/java/com/smsrelay/mvp/SecureChannel.kt
package com.smsrelay.mvp

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

object SecureChannel {
    const val PROTOCOL_VERSION = 2
    private const val AUTH_CONTEXT = "Tang/auth/v2"
    private const val CLIENT_TO_SERVER_CONTEXT = "Tang/client-to-server/v2"
    private const val SERVER_TO_CLIENT_CONTEXT = "Tang/server-to-client/v2"
    private const val FRAME_CONTEXT = "Tang/frame/v2/"
    private const val NONCE_BYTES = 12
    private const val TAG_BITS = 128
    private val random = SecureRandom()

    data class DirectionalKeys(
        val clientToServer: ByteArray,
        val serverToClient: ByteArray
    )

    data class EncryptedFrame(
        val sequence: Long,
        val nonceBase64: String,
        val ciphertextBase64: String
    )

    fun randomNonceBase64(bytes: Int = 32): String {
        val value = ByteArray(bytes)
        random.nextBytes(value)
        return Base64.getEncoder().encodeToString(value)
    }

    fun authProofBase64(
        secret: String,
        clientNonceBase64: String,
        challengeBase64: String,
        device: String,
        appVersion: String
    ): String {
        val proof = hmacSha256(
            key = secret.toByteArray(StandardCharsets.UTF_8),
            data = authInput(clientNonceBase64, challengeBase64, device, appVersion)
        )
        return Base64.getEncoder().encodeToString(proof)
    }

    fun proofMatches(expectedBase64: String, providedBase64: String): Boolean {
        val expected = decodeBase64(expectedBase64) ?: return false
        val provided = decodeBase64(providedBase64) ?: return false
        return MessageDigest.isEqual(expected, provided)
    }

    fun deriveKeys(
        secret: String,
        clientNonceBase64: String,
        challengeBase64: String
    ): DirectionalKeys {
        val clientNonce = requireNotNull(decodeBase64(clientNonceBase64)) { "invalid client nonce" }
        val challenge = requireNotNull(decodeBase64(challengeBase64)) { "invalid challenge" }
        require(clientNonce.size == 32) { "client nonce must be 32 bytes" }
        require(challenge.size == 32) { "challenge must be 32 bytes" }
        val salt = clientNonce + challenge
        val secretBytes = secret.toByteArray(StandardCharsets.UTF_8)
        return DirectionalKeys(
            clientToServer = hkdfSha256(secretBytes, salt, CLIENT_TO_SERVER_CONTEXT.toByteArray(), 32),
            serverToClient = hkdfSha256(secretBytes, salt, SERVER_TO_CLIENT_CONTEXT.toByteArray(), 32)
        )
    }

    class Session(
        private val sendKey: ByteArray,
        private val receiveKey: ByteArray
    ) {
        private var sendSequence = 0L
        private var receiveSequence = 0L

        @Synchronized
        fun encrypt(plaintext: String): EncryptedFrame {
            val nonce = ByteArray(NONCE_BYTES)
            random.nextBytes(nonce)
            return encrypt(plaintext, nonce)
        }

        @Synchronized
        internal fun encrypt(plaintext: String, nonce: ByteArray): EncryptedFrame {
            require(nonce.size == NONCE_BYTES) { "nonce must be 12 bytes" }
            val sequence = Math.addExact(sendSequence, 1L)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(
                Cipher.ENCRYPT_MODE,
                SecretKeySpec(sendKey, "AES"),
                GCMParameterSpec(TAG_BITS, nonce)
            )
            cipher.updateAAD(frameAad(sequence))
            val ciphertext = cipher.doFinal(plaintext.toByteArray(StandardCharsets.UTF_8))
            sendSequence = sequence
            return EncryptedFrame(
                sequence = sequence,
                nonceBase64 = Base64.getEncoder().encodeToString(nonce),
                ciphertextBase64 = Base64.getEncoder().encodeToString(ciphertext)
            )
        }

        @Synchronized
        fun decrypt(frame: EncryptedFrame): String? {
            if (frame.sequence != receiveSequence + 1) {
                return null
            }
            val nonce = decodeBase64(frame.nonceBase64) ?: return null
            val ciphertext = decodeBase64(frame.ciphertextBase64) ?: return null
            if (nonce.size != NONCE_BYTES || ciphertext.size < TAG_BITS / 8) {
                return null
            }
            val plaintext = runCatching {
                val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                cipher.init(
                    Cipher.DECRYPT_MODE,
                    SecretKeySpec(receiveKey, "AES"),
                    GCMParameterSpec(TAG_BITS, nonce)
                )
                cipher.updateAAD(frameAad(frame.sequence))
                cipher.doFinal(ciphertext)
            }.getOrNull() ?: return null
            receiveSequence = frame.sequence
            return plaintext.toString(StandardCharsets.UTF_8)
        }
    }

    private fun authInput(
        clientNonceBase64: String,
        challengeBase64: String,
        device: String,
        appVersion: String
    ): ByteArray {
        val encoder = Base64.getEncoder()
        val deviceBase64 = encoder.encodeToString(device.toByteArray(StandardCharsets.UTF_8))
        val versionBase64 = encoder.encodeToString(appVersion.toByteArray(StandardCharsets.UTF_8))
        return listOf(
            AUTH_CONTEXT,
            clientNonceBase64,
            challengeBase64,
            deviceBase64,
            versionBase64
        ).joinToString("\n").toByteArray(StandardCharsets.UTF_8)
    }

    private fun frameAad(sequence: Long): ByteArray {
        return "$FRAME_CONTEXT$sequence".toByteArray(StandardCharsets.UTF_8)
    }

    private fun decodeBase64(value: String): ByteArray? {
        return runCatching { Base64.getDecoder().decode(value) }.getOrNull()
    }

    private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    private fun hkdfSha256(
        inputKeyMaterial: ByteArray,
        salt: ByteArray,
        info: ByteArray,
        outputLength: Int
    ): ByteArray {
        require(outputLength in 1..32) { "only one SHA-256 HKDF block is supported" }
        val pseudoRandomKey = hmacSha256(salt, inputKeyMaterial)
        val block = hmacSha256(pseudoRandomKey, info + byteArrayOf(1))
        return block.copyOf(outputLength)
    }
}

File: macos-menubar/Sources/SmsRelayMenuBar/SecureChannel.swift
import CryptoKit
import Foundation
import Security

enum SecureChannel {
    static let protocolVersion = 2
    private static let authContext = "Tang/auth/v2"
    private static let clientToServerContext = "Tang/client-to-server/v2"
    private static let serverToClientContext = "Tang/server-to-client/v2"
    private static let frameContext = "Tang/frame/v2/"

    struct DirectionalKeys {
        let clientToServer: Data
        let serverToClient: Data
    }

    struct EncryptedFrame {
        let sequence: UInt64
        let nonceBase64: String
        let ciphertextBase64: String
    }

    static func randomNonceBase64(bytes: Int = 32) -> String {
        var data = Data(count: bytes)
        let status = data.withUnsafeMutableBytes { pointer in
            SecRandomCopyBytes(kSecRandomDefault, bytes, pointer.baseAddress!)
        }
        precondition(status == errSecSuccess)
        return data.base64EncodedString()
    }

    static func authProofBase64(
        secret: String,
        clientNonceBase64: String,
        challengeBase64: String,
        device: String,
        appVersion: String
    ) -> String {
        let proof = hmacSha256(
            key: Data(secret.utf8),
            data: authInput(
                clientNonceBase64: clientNonceBase64,
                challengeBase64: challengeBase64,
                device: device,
                appVersion: appVersion
            )
        )
        return proof.base64EncodedString()
    }

    static func proofMatches(expectedBase64: String, providedBase64: String) -> Bool {
        guard let expected = Data(base64Encoded: expectedBase64),
              let provided = Data(base64Encoded: providedBase64),
              expected.count == provided.count else {
            return false
        }
        return zip(expected, provided).reduce(UInt8(0)) { result, pair in
            result | (pair.0 ^ pair.1)
        } == 0
    }

    static func deriveKeys(
        secret: String,
        clientNonceBase64: String,
        challengeBase64: String
    ) throws -> DirectionalKeys {
        guard let clientNonce = Data(base64Encoded: clientNonceBase64), clientNonce.count == 32,
              let challenge = Data(base64Encoded: challengeBase64), challenge.count == 32 else {
            throw SecureChannelError.invalidNonce
        }
        var salt = Data()
        salt.append(clientNonce)
        salt.append(challenge)
        let inputKeyMaterial = Data(secret.utf8)
        return DirectionalKeys(
            clientToServer: hkdfSha256(
                inputKeyMaterial: inputKeyMaterial,
                salt: salt,
                info: Data(clientToServerContext.utf8),
                outputLength: 32
            ),
            serverToClient: hkdfSha256(
                inputKeyMaterial: inputKeyMaterial,
                salt: salt,
                info: Data(serverToClientContext.utf8),
                outputLength: 32
            )
        )
    }

    struct Session {
        private let sendKey: SymmetricKey
        private let receiveKey: SymmetricKey
        private var sendSequence: UInt64 = 0
        private var receiveSequence: UInt64 = 0

        init(sendKey: Data, receiveKey: Data) {
            self.sendKey = SymmetricKey(data: sendKey)
            self.receiveKey = SymmetricKey(data: receiveKey)
        }

        mutating func encrypt(_ plaintext: String) throws -> EncryptedFrame {
            let nonceData = Data((0..<12).map { _ in UInt8.random(in: .min ... .max) })
            return try encrypt(plaintext, nonce: nonceData)
        }

        mutating func encrypt(_ plaintext: String, nonce: Data) throws -> EncryptedFrame {
            guard nonce.count == 12 else {
                throw SecureChannelError.invalidNonce
            }
            let sequence = sendSequence.addingReportingOverflow(1)
            guard !sequence.overflow else {
                throw SecureChannelError.sequenceOverflow
            }
            let sealed = try AES.GCM.seal(
                Data(plaintext.utf8),
                using: sendKey,
                nonce: try AES.GCM.Nonce(data: nonce),
                authenticating: frameAad(sequence.partialValue)
            )
            var ciphertext = sealed.ciphertext
            ciphertext.append(sealed.tag)
            sendSequence = sequence.partialValue
            return EncryptedFrame(
                sequence: sequence.partialValue,
                nonceBase64: nonce.base64EncodedString(),
                ciphertextBase64: ciphertext.base64EncodedString()
            )
        }

        mutating func decrypt(_ frame: EncryptedFrame) -> String? {
            guard frame.sequence == receiveSequence + 1,
                  let nonceData = Data(base64Encoded: frame.nonceBase64), nonceData.count == 12,
                  let ciphertextAndTag = Data(base64Encoded: frame.ciphertextBase64),
                  ciphertextAndTag.count >= 16 else {
                return nil
            }
            let ciphertext = ciphertextAndTag.dropLast(16)
            let tag = ciphertextAndTag.suffix(16)
            guard let sealedBox = try? AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonceData),
                ciphertext: ciphertext,
                tag: tag
            ),
            let plaintext = try? AES.GCM.open(
                sealedBox,
                using: receiveKey,
                authenticating: SecureChannel.frameAad(frame.sequence)
            ),
            let value = String(data: plaintext, encoding: .utf8) else {
                return nil
            }
            receiveSequence = frame.sequence
            return value
        }
    }

    enum SecureChannelError: Error {
        case invalidNonce
        case sequenceOverflow
    }

    private static func authInput(
        clientNonceBase64: String,
        challengeBase64: String,
        device: String,
        appVersion: String
    ) -> Data {
        let fields = [
            authContext,
            clientNonceBase64,
            challengeBase64,
            Data(device.utf8).base64EncodedString(),
            Data(appVersion.utf8).base64EncodedString()
        ]
        return Data(fields.joined(separator: "\n").utf8)
    }

    private static func frameAad(_ sequence: UInt64) -> Data {
        Data("\(frameContext)\(sequence)".utf8)
    }

    private static func hmacSha256(key: Data, data: Data) -> Data {
        let code = HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key))
        return Data(code)
    }

    private static func hkdfSha256(
        inputKeyMaterial: Data,
        salt: Data,
        info: Data,
        outputLength: Int
    ) -> Data {
        precondition((1...32).contains(outputLength))
        let pseudoRandomKey = hmacSha256(key: salt, data: inputKeyMaterial)
        var blockInput = info
        blockInput.append(1)
        return Data(hmacSha256(key: pseudoRandomKey, data: blockInput).prefix(outputLength))
    }
}
